File: training-data/python/longest_increasing_subsequence.py
import bisect
from typing import List


def length_of_lis(nums: List[int]) -> int:
    tails: List[int] = []
    for num in nums:
        idx = bisect.bisect_left(tails, num)
        if idx == len(tails):
            tails.append(num)
        else:
            tails[idx] = num
    return len(tails)


if __name__ == "__main__":
    print(length_of_lis([10, 9, 2, 5, 3, 7, 101, 18]))

File: training-data/csharp/LongestIncreasingSubsequence.cs
using System;
using System.Collections.Generic;

class LongestIncreasingSubsequence
{
    static int LengthOfLIS(int[] nums)
    {
        var tails = new List<int>();
        foreach (int num in nums)
        {
            int idx = tails.BinarySearch(num);
            if (idx < 0) idx = ~idx;
            if (idx == tails.Count) tails.Add(num);
            else tails[idx] = num;
        }
        return tails.Count;
    }

    static void Main()
    {
        Console.WriteLine(LengthOfLIS(new[] { 10, 9, 2, 5, 3, 7, 101, 18 }));
    }
}
